gateway-server/src/main/java/org/apache/knox/gateway/webshell/WebshellWebSocketAdapter.java

```diff
@@ -0,0 +1,161 @@
+ */
+package org.apache.knox.gateway.webshell;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.ExecutorService;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.knox.gateway.config.GatewayConfig;
+import org.apache.knox.gateway.services.GatewayServices;
+import org.apache.knox.gateway.services.security.token.UnknownTokenException;
+import org.apache.knox.gateway.websockets.ProxyWebSocketAdapter;
+import org.eclipse.jetty.websocket.api.Session;
+import org.eclipse.jetty.websocket.servlet.ServletUpgradeRequest;
+
+public class WebshellWebSocketAdapter extends ProxyWebSocketAdapter  {
+    private Session session;
+    private ConnectionInfo connectionInfo;
+    private String username;
+
+    public WebshellWebSocketAdapter(ServletUpgradeRequest req,  ExecutorService pool, GatewayConfig config, GatewayServices services) {
+        super(null, pool, null, config);
+        // todo: may not need to get username from request URI, username is contained in JWT token
+        if (config.isWebShellEnabled()) {
+            username = req.getRequestURI().getRawQuery();
+            try {
+                WebShellTokenUtils.validateJWT(req, services, config);
+            } catch (final UnknownTokenException e) {
+                LOG.onError("no valid token found");
+                // todo: how to handle validation error?
+                throw new RuntimeException(e);
+            }
+            this.connectionInfo = new ProcessConnectionInfo(username);
+        }
+        else {
+            throw new RuntimeException("webshell not enabled");
+        }
+    }
+
+    @Override
+    public void onWebSocketConnect(final Session session) {
+        this.session = session;
+        LOG.debugLog("websocket connected.");
+        connectionInfo.connect();
+        pool.execute(new Runnable() {
+            @Override
+            public void run() {
+                blockingReadFromHost();
+            }
+        });
+    }
+
+    // this function will block, should be run in an asynchronous thread
+    private void blockingReadFromHost(){
+        LOG.debugLog("start listening to bash process");
+        byte[] buffer = new byte[1024];
+        int bytesRead;
+        try {
+            // todo: maybe change to non-blocking read using java.nio
+            while ((bytesRead = connectionInfo.getInputStream().read(buffer)) != -1) {
+                transToClient(new String(buffer, 0, bytesRead, StandardCharsets.UTF_8));
+            }
+        } catch (IOException e){
+            transToClient("bash process terminated unexpectedly");
+        } finally {
+            cleanupOnError("bash process terminated unexpectedly");
+        }
+    }
+
+    @SuppressWarnings("PMD.DoNotUseThreads")
+    @Override
+    public void onWebSocketText(final String message) {
+        LOG.debugLog("received message "+ message);
+
```

Comment:
Let's check for token expiry here.

Reply:
> Let's check for token expiry here.
does each message carry the JWT with it?